### PR TITLE
Added support for Homebrew release

### PR DESCRIPTION
**Pull request by sabre1041**
Added support for release to [Homebrew Tap](https://github.com/redhat-cop/homebrew-redhat-cop)

@bserdar Need to sync up with you to set repository secrets to enable pushing
@rpelisse Feel free to review

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -30,3 +30,38 @@ jobs:
...
+          BREW_USER: ${{ secrets.BREW_USER }}
+          BREW_TOKEN: ${{ secrets.BREW_TOKEN }}
+        run: |
+          set -e
+
+          JCLIFF_TAG=${{ steps.get_version.outputs.TAG_VERSION }}
+          HOMEBREW_TAP_REPO="${HOMEBREW_TAP_REPO:-redhat-cop/homebrew-redhat-cop}"
+          REPO_NAME=$(basename $HOMEBREW_TAP_REPO)
+
+          SHA256=`sha256sum target/jcliff-${JCLIFF_TAG:1}-dist.tar.gz | cut -f1 -d ' '`
```
Comment (rpelisse) [inline .github/workflows/release.yml:45]: @sabre1041 you are way too young to use `` instead of $() ;)

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -30,3 +30,38 @@ jobs:
...
           body: "Release ${{ steps.get_version.outputs.TAG_VERSION }}"
+      - name: Homebrew Release
+        env:
+          HOMEBREW_TAP_REPO: ${{ secrets.HOMEBREW_REPOSITORY }}
+          BREW_USER: ${{ secrets.BREW_USER }}
+          BREW_TOKEN: ${{ secrets.BREW_TOKEN }}
+        run: |
+          set -e
+
+          JCLIFF_TAG=${{ steps.get_version.outputs.TAG_VERSION }}
```
Comment (rpelisse) [inline .github/workflows/release.yml:41]: I would prefix all those vars with readonly (or use lowercase)

Comment (rpelisse) [reply .github/workflows/release.yml:41]: Also surround the var with double-quote

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -30,3 +30,38 @@ jobs:
...
+          HOMEBREW_TAP_REPO="${HOMEBREW_TAP_REPO:-redhat-cop/homebrew-redhat-cop}"
+          REPO_NAME=$(basename $HOMEBREW_TAP_REPO)
```
Comment (rpelisse) [inline .github/workflows/release.yml:43]: => "${HOMEBREW_TAP_REPO}"

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -30,3 +30,38 @@ jobs:
...
+
+          SHA256=`sha256sum target/jcliff-${JCLIFF_TAG:1}-dist.tar.gz | cut -f1 -d ' '`
+          URL="https://github.com/bserdar/jcliff/releases/download/${JCLIFF_TAG}/jcliff-${JCLIFF_TAG:1}-dist.tar.gz"
+
+          rm -rf $REPO_NAME
```
Comment (rpelisse) [inline .github/workflows/release.yml:48]: => "${REPO_NAME}"

especially here. Imagine:

REPO_NAME='a_file /' => boom root is erased!

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -30,3 +30,38 @@ jobs:
...
+
+          git clone https://github.com/${HOMEBREW_TAP_REPO}.git
```
Comment (rpelisse) [inline .github/workflows/release.yml:50]: I would add REPO_URL var and also add "" around the var substitution.

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -30,3 +30,38 @@ jobs:
...
+
+          sed -i -e 's|^  url.*|  url "'"$URL"'"|g' $REPO_NAME/Formula/jcliff.rb
```
Comment (rpelisse) [inline .github/workflows/release.yml:52]: => "${REPO_NAME}/Formula/jcliff.rb"

Comment (rpelisse) [reply .github/workflows/release.yml:52]: => "${URL}"

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -30,3 +30,38 @@ jobs:
...
+          sed -i -e 's|^  sha256.*|  sha256 "'"$SHA256"'"|g' $REPO_NAME/Formula/jcliff.rb
```
Comment (rpelisse) [inline .github/workflows/release.yml:53]: => "${SHA256}"

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -30,3 +30,38 @@ jobs:
...
+
+          cd $(basename $HOMEBREW_TAP_REPO)
```
Comment (rpelisse) [inline .github/workflows/release.yml:55]: "${HOMEBREW_TAP_REPO}"

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -30,3 +30,38 @@ jobs:
...
+
+          git config user.name "Andrew Block"
```
Comment (rpelisse) [inline .github/workflows/release.yml:57]: Username, email => should be in vars.

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -30,3 +30,38 @@ jobs:
...
+          git config user.email "andy.block@gmail.com"
+
+          git add Formula/jcliff.rb
+          git commit -m "jcliff ${JCLIFF_TAG} release"
+
+          if [[ ! -z $BREW_USER && ! -z $BREW_TOKEN ]]; then
```
Comment (rpelisse) [inline .github/workflows/release.yml:63]: "${BREW_USER}" and "${BREW_TOKEN}"

File: .github/workflows/release.yml [hunk 11]
```diff
@@ -30,3 +30,38 @@ jobs:
...
+            git push https://${BREW_USER}:${BREW_GITHUB_TOKEN}@github.com/${HOMEBREW_TAP_REPO}.git
```
Comment (rpelisse) [inline .github/workflows/release.yml:64]: => add missing double quote : "https://${BREW_USER}:${BREW_GITHUB_TOKEN}@github.com/${HOMEBREW_TAP_REPO}.git"

File: .github/workflows/release.yml [hunk 12]
```diff
@@ -30,3 +30,38 @@ jobs:
...
+          else
+            echo "BREW_USER and BREW_TOKEN variables not set. Skipping push..."
```
Comment (rpelisse) [inline .github/workflows/release.yml:66]: echo "BREW_USER and BREW_TOKEN variables not set or are invalid:"
echo "BREW_USER: ${BREW_USER}"
echo "BREW_TOKEN: ${BREW_TOKEN}"
echo "Skipping push..."

File: .github/workflows/release.yml [hunk 13]
```diff
@@ -30,3 +30,38 @@ jobs:
             target/*.tar.gz
             target/rpm/jcliff/RPMS/noarch/*.rpm
           body: "Release ${{ steps.get_version.outputs.TAG_VERSION }}"
+      - name: Homebrew Release
+        env:
+          HOMEBREW_TAP_REPO: ${{ secrets.HOMEBREW_REPOSITORY }}
+          BREW_USER: ${{ secrets.BREW_USER }}
+          BREW_TOKEN: ${{ secrets.BREW_TOKEN }}
+        run: |
```
Comment (rpelisse) [inline .github/workflows/release.yml:38]: @sabre1041 anyway to externalise this bash script into a separate file? It would be cleaner IMH.

**Review by rpelisse — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml (hunk 13 above)
Comment (sabre1041) [reply .github/workflows/release.yml:38]: Sure. Just need to know where in the repository it should go

Comment (rpelisse) [reply .github/workflows/release.yml:38]: I generally put this kind of scripts into src/build/bash.

**Comment by sabre1041** [comment]
@rpelisse Addressed your concerns. The only area I did not address was for the Git URL. I dont see us going off of GitHub any time soon. However, we might want to look to migrating the Homebrew repository from the `redhat-cop` org to `ansible-middleware`.

**Comment by bserdar** [comment]
@sabre1041 what do you need re: repository secrets?

File: src/main/scripts/homebrew-release.sh [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+readonly SHA256=$(sha256sum target/jcliff-${JCLIFF_TAG:1}-dist.tar.gz | cut -f1 -d ' ')
+readonly URL="https://github.com/bserdar/jcliff/releases/download/${JCLIFF_TAG}/jcliff-${JCLIFF_TAG:1}-dist.tar.gz"
+
+rm -rf "${REPO_NAME}"
+
+git clone https://github.com/${HOMEBREW_TAP_REPO}.git
+
+sed -i -e 's|^  url.*|  url "'"$URL"'"|g' ${REPO_NAME}/Formula/jcliff.rb
+sed -i -e 's|^  sha256.*|  sha256 "'"$SHA256"'"|g' ${REPO_NAME}/Formula/jcliff.rb
```
[the remark this answers is not in the document]
Comment (rpelisse) [reply src/main/scripts/homebrew-release.sh:15]: It's probably overdoing it, but know that you can check the status of each commands in pipe using [the variable PIPESTATUS](https://tldp.org/LDP/abs/html/internalvariables.html#PIPESTATUSREF).

**Review by rpelisse — CHANGES_REQUESTED** [review]
With all the missing overrides in place, we can further and have Bats unit test setup to validate that this script behave properly. (but it's probably overdoing it a bit)

File: src/main/scripts/homebrew-release.sh [hunk 15]
```diff
@@ -0,0 +1,30 @@
+set -e
+
+readonly JCLIFF_TAG="${{ steps.get_version.outputs.TAG_VERSION }}"
```
[the remark this answers is not in the document]
Comment (rpelisse) [reply src/main/scripts/homebrew-release.sh:3]: Of course... I really need to learn about Github Actions :) Please ignore.

File: src/main/scripts/homebrew-release.sh [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+rm -rf "${REPO_NAME}"
+
+git clone https://github.com/${HOMEBREW_TAP_REPO}.git
+
+sed -i -e 's|^  url.*|  url "'"$URL"'"|g' ${REPO_NAME}/Formula/jcliff.rb
+sed -i -e 's|^  sha256.*|  sha256 "'"$SHA256"'"|g' ${REPO_NAME}/Formula/jcliff.rb
+
+cd $(basename "$HOMEBREW_TAP_REPO")
+
+git config user.name "${BREW_GITHUB_USER:-Andrew Block}"
```
[the remark this answers is not in the document]
Comment (rpelisse) [reply src/main/scripts/homebrew-release.sh:23]: @sabre1041 What about this?

File: src/main/scripts/homebrew-release.sh [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+readonly URL="https://github.com/bserdar/jcliff/releases/download/${JCLIFF_TAG}/jcliff-${JCLIFF_TAG:1}-dist.tar.gz"
+
+rm -rf "${REPO_NAME}"
+
+git clone https://github.com/${HOMEBREW_TAP_REPO}.git
+
+sed -i -e 's|^  url.*|  url "'"$URL"'"|g' ${REPO_NAME}/Formula/jcliff.rb
+sed -i -e 's|^  sha256.*|  sha256 "'"$SHA256"'"|g' ${REPO_NAME}/Formula/jcliff.rb
+
+cd $(basename "$HOMEBREW_TAP_REPO")
```
[the remark this answers is not in the document]
Comment (rpelisse) [reply src/main/scripts/homebrew-release.sh:21]: @sabre1041 I think you missed this :)

**Review by rpelisse — APPROVED** [review]
There is a few things that can still be fixed up, but apart from those, I want to say good to go!

**Comment by sabre1041** [comment]
@rpelisse fixed. hopefully last one

**Comment by rpelisse** [comment]
@sabre1041 yes, you have indulged my bash-extremism enough :) !

**Pull request merged**